File: app/src/main/java/ru/p3tr0vich/fuel/ActivityDialog.java
package ru.p3tr0vich.fuel;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class ActivityDialog extends AppCompatActivity {

    private static final String DIALOG = "DIALOG";


    @Retention(RetentionPolicy.SOURCE)
    @IntDef({DIALOG_SMS_TEXT_PATTERN})
    public @interface Dialog {
    }

    public static final int DIALOG_SMS_TEXT_PATTERN = 0;

    public interface ActivityDialogFragment {
        String getTitle();

        boolean onSaveClicked();
    }

    public static void start(@NonNull Activity parent, @Dialog int dialog) {
        parent.startActivity(new Intent(parent, ActivityDialog.class).putExtra(DIALOG, dialog));
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_dialog);

        Toolbar toolbarDialog = (Toolbar) findViewById(R.id.toolbarDialog);
        setSupportActionBar(toolbarDialog);

        assert toolbarDialog != null;

        toolbarDialog.setNavigationIcon(R.drawable.ic_close);

        toolbarDialog.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setResult(RESULT_CANCELED, null);
                finish();
            }
        });

        if (savedInstanceState == null) {
            Fragment fragment;

            switch (getIntent().getIntExtra(DIALOG, -1)) {
                case DIALOG_SMS_TEXT_PATTERN:
                    fragment = FragmentActivityDialogSMSTextPattern.newInstance();
                    break;
                default:
                    throw new IllegalArgumentException("Unknown dialog type");
            }

            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.contentFrame, fragment, null)
                    .setTransition(FragmentTransaction.TRANSIT_NONE)
                    .commit();
        }
    }

    @Override
    public void onAttachFragment(Fragment fragment) {
        setTitle(((ActivityDialogFragment) fragment).getTitle());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_action_save, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        ActivityDialogFragment fragment =
                (ActivityDialogFragment) getSupportFragmentManager().findFragmentById(R.id.contentFrame);

        if (fragment.onSaveClicked()) {
            setResult(Activity.RESULT_OK);
            finish();
        }

        return true;
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/activities/ActivityDialog.kt
package ru.p3tr0vich.fuel.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.annotation.IntDef
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentOnAttachListener
import androidx.fragment.app.FragmentTransaction
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.fragments.FragmentActivityDialogSMSTextPattern

class ActivityDialog : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_dialog)

        val toolbarDialog = findViewById<Toolbar>(R.id.toolbar_dialog)

        setSupportActionBar(toolbarDialog)

        toolbarDialog?.setNavigationIcon(R.drawable.ic_close)

        toolbarDialog?.setNavigationOnClickListener {
            setResult(Activity.RESULT_CANCELED, null)
            finish()
        }

        supportFragmentManager.addFragmentOnAttachListener(object : FragmentOnAttachListener {
            override fun onAttachFragment(fragmentManager: FragmentManager, fragment: Fragment) {
                title = (fragment as ActivityDialogFragment).title
            }
        })

        if (savedInstanceState == null) {
            val fragment: Fragment

            when (intent.getIntExtra(DIALOG, -1)) {
                DIALOG_SMS_TEXT_PATTERN -> fragment =
                    FragmentActivityDialogSMSTextPattern.newInstance()

                else -> throw IllegalArgumentException("Unknown dialog type")
            }

            supportFragmentManager.beginTransaction()
                .replace(R.id.content_frame, fragment, null)
                .setTransition(FragmentTransaction.TRANSIT_NONE)
                .commit()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_action_save, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val fragment =
            supportFragmentManager.findFragmentById(R.id.content_frame) as ActivityDialogFragment?

        if (fragment != null && fragment.onSaveClicked()) {
            setResult(Activity.RESULT_OK)
            finish()
        }

        return true
    }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(DIALOG_SMS_TEXT_PATTERN)
    annotation class Dialog

    interface ActivityDialogFragment {
        val title: String

        fun onSaveClicked(): Boolean
    }

    companion object {
        private const val DIALOG = "DIALOG"

        const val DIALOG_SMS_TEXT_PATTERN = 0

        @JvmStatic
        fun start(parent: Activity, @Dialog dialog: Int) {
            parent.startActivity(
                Intent(parent, ActivityDialog::class.java).putExtra(
                    DIALOG,
                    dialog
                )
            )
        }
    }
}